SpatialGDK/Source/SpatialGDKFunctionalTests/Private/SpatialCleanupConnectionTest.cpp

```diff
@@ -0,0 +1,93 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+#include "SpatialCleanupConnectionTest.h"
+
+#include "SpatialFunctionalTestFlowController.h"
+
+ASpatialCleanupConnectionTest::ASpatialCleanupConnectionTest()
+{
+	Author = "Simon Sarginson";
+	Description = TEXT("Tests if clients connections are cleaned up when leaving interest region of a non authorative server");
+
+	Server1Position = FVector(-500.0f, -500.0f, 50.0f);
+	Server1PositionAndInInterestBorderServer2 = FVector(-50.0f, -50.0f, 50.0f);
+	Server2Position = FVector(500.0f, -500.0f, 50.0f);
+
+	SetNumRequiredClients(1);
+}
+
+void ASpatialCleanupConnectionTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	AddStep(TEXT("Spawn player on server 1"), FWorkerDefinition::Server(1), nullptr, [this]() {
+		USpatialNetDriver* Driver = Cast<USpatialNetDriver>(GetNetDriver());
+		AssertIsValid(Driver, TEXT("Test is exclusive to using SpatialNetDriver"));
+		ASpatialFunctionalTestFlowController* ClientOneFlowController = GetFlowController(ESpatialFunctionalTestWorkerType::Client, 1);
+		PlayerController = Cast<APlayerController>(ClientOneFlowController->GetOwner());
+		AssertIsValid(PlayerController, TEXT("Must have valid PlayerController for test"));
+		DefaultPawn = PlayerController->GetPawn();
+		PlayerController->UnPossess();
+		SpawnedPawn = GetWorld()->SpawnActor<ATestMovementCharacter>(Server1Position, FRotator::ZeroRotator, FActorSpawnParameters());
+		RegisterAutoDestroyActor(SpawnedPawn);
+		PlayerController->Possess(SpawnedPawn);
+
+		AssertEqual_Int(Driver->ClientConnections.Num(), 2, TEXT("Spawn expected 2 connections (base + PC)"));
+		FinishStep();
+	});
+
+	AddStep(
+		TEXT("Post spawn check connections on server 2"), FWorkerDefinition::Server(2), nullptr, nullptr,
+		[this](float delta) {
+			USpatialNetDriver* Driver = Cast<USpatialNetDriver>(GetNetDriver());
+			AssertIsValid(Driver, TEXT("Test is exclusive to using SpatialNetDriver"));
+			RequireEqual_Int(Driver->ClientConnections.Num(), 1, TEXT("Spawn expected 1 connection (base only)"));
+			FinishStep();
```

Comment:
If I remember correctly, RequireEqual_Int is doing the FinishStep for you when the condition passes.
What you are doing looks like it is skipping it.

Reply:
That's not the case. Requires work pretty much the same as asserts except they don't immediately stop execution on failure. Instead, the status of the requires will be reported on `FinishStep`.
In this case, there wouldn't really be a difference between using Require or Assert since nothing happens after the Require.

Reply:
As far as I understand what @mironec explained to me its still needed, RequireEqual prevents FinishStep from being called.

Reply:
We need the require because we need to wait for the async operation for removing the connection on server2 to complete

Reply:
That doesn't match what I see in the code for `RequireEqual` and `FinishStep`. As far as I can see, `RequireEqual` will immediately evaluate the condition and add it to the list. `FinishStep` will then check that list and report any failures. At no point is there async waiting.

Reply:
Ah yes, my bad, I misunderstood how requires worked. Your usage looks correct.